### PR TITLE
OPENTOK-46715: iOS Custom Audio Driver sync with SDK implementation

### PR DESCRIPTION
**Pull request by IGitGotIt**
Sync of sample code  with SDK code

### CODE REVIEW

File: Custom-Audio-Driver/External-Audio-Device/OTDefaultAudioDevice.m [hunk 1]
```diff
@@ -434,6 +434,8 @@ - (void)freeupAudioBuffers
 }
 - (void) setupAudioSession
 {
+    if(isAudioSessionSetup) return;
```
Comment (v-kpheng) [inline Custom-Audio-Driver/External-Audio-Device/OTDefaultAudioDevice.m:437]: Nit: space after `if` would be nice 🤞

File: Custom-Audio-Driver/External-Audio-Device/OTDefaultAudioDevice.m [hunk 2]
```diff
@@ -476,31 +478,38 @@ - (void) setupAudioSession
     
     if (error)
         OT_AUDIO_DEBUG(@"Audiosession setActive %@",error);
-    
-    [self setBluetoothAsPrefferedInputDevice];
+
+    if (@available(iOS 15, *)) {
+        // do nothing refer comments for self.setBluetoothAsPrefferedInputDevice method.
```
Comment (v-kpheng) [inline Custom-Audio-Driver/External-Audio-Device/OTDefaultAudioDevice.m:483]: Why not log this instead?  A comment may be easy to miss.

File: Custom-Audio-Driver/External-Audio-Device/OTDefaultAudioDevice.m [hunk 3]
```diff
@@ -476,31 +478,38 @@ - (void) setupAudioSession
...
-    {
-        if ([bluetoothRoutes containsObject:route.portType])
-        {
-            [[AVAudioSession sharedInstance] setPreferredInput:route
-                                                         error:nil];
-            break;
-        }
+- (void)setBluetoothAsPrefferedInputDevice {
+  // Apple's Bug(???) : Before iOS 15 AVAudioSessionInterruptionTypeEnded notification would
+  // not be called for bluetooth if we dont set bluetooth as preffered input as
```
Comment (v-kpheng) [inline Custom-Audio-Driver/External-Audio-Device/OTDefaultAudioDevice.m:492]: Nit.  Misspelled preferred.

File: Custom-Audio-Driver/External-Audio-Device/OTDefaultAudioDevice.m [hunk 4]
```diff
@@ -476,31 +478,38 @@ - (void) setupAudioSession
...
+  // in setupAudioSession. In iOS 15 starting a session with BT and then disconnecting it
+  // would cause the camera to freeze for some reason (an Apple Bug again ??)
+  //
+  // This method is also called on AVAudioSessionInterruptionTypeEnded, because BT audio
+  // routing would be lost, if you recieved a phone call and ended it, while
```
Comment (v-kpheng) [inline Custom-Audio-Driver/External-Audio-Device/OTDefaultAudioDevice.m:497]: Nit.  Misspelled received.

File: Custom-Audio-Driver/External-Audio-Device/OTDefaultAudioDevice.m [hunk 5]
```diff
@@ -476,31 +478,38 @@ - (void) setupAudioSession
...
+  // using BT with OT
+  //
+  // Should work for non bluetooth routes/ports too. This makes both input
+  // and output to bluetooth device if available.
+  NSArray *bluetoothRoutes = @[
+    AVAudioSessionPortBluetoothA2DP, AVAudioSessionPortBluetoothLE,
+    AVAudioSessionPortBluetoothHFP
+  ];
+  NSArray *routes = [[AVAudioSession sharedInstance] availableInputs];
```
Comment (v-kpheng) [inline Custom-Audio-Driver/External-Audio-Device/OTDefaultAudioDevice.m:506]: Looks like this sets up BT routes

File: Custom-Audio-Driver/External-Audio-Device/OTDefaultAudioDevice.m [hunk 6]
```diff
@@ -543,6 +552,11 @@ - (void) handleInterruptionEvent:(NSInteger) interruptionType
             case AVAudioSessionInterruptionTypeEnded:
             {
                 OT_AUDIO_DEBUG(@"AVAudioSessionInterruptionTypeEnded");
+                if (@available(iOS 15, *)) {
+                    [self setBluetoothAsPrefferedInputDevice];
+                } else {
+                    // do nothing refer comments for self.setBluetoothAsPrefferedInputDevice method.
```
Comment (v-kpheng) [inline Custom-Audio-Driver/External-Audio-Device/OTDefaultAudioDevice.m:558]: Same thing here.  How do you feel about logging this, instead, to make it more difficult to miss?

File: Custom-Audio-Driver/External-Audio-Device/OTDefaultAudioDevice.m [hunk 7]
```diff
@@ -1102,6 +1117,22 @@ - (BOOL)setupAudioUnit:(AudioUnit *)voice_unit playout:(BOOL)isPlayout;
     
     // Initialize the Voice-Processing I/O unit instance.
     result = AudioUnitInitialize(*voice_unit);
+
+    // This patch is pickedup from WebRTC audio implementation and
```
Comment (v-kpheng) [inline Custom-Audio-Driver/External-Audio-Device/OTDefaultAudioDevice.m:1121]: "pickedup" -> "picked up"

**Review by v-kpheng — APPROVED** [review]
LGTM! 💪 🚀

File: Custom-Audio-Driver/External-Audio-Device/OTDefaultAudioDevice.m (hunk 2 above)
Comment (IGitGotIt) [reply Custom-Audio-Driver/External-Audio-Device/OTDefaultAudioDevice.m:483]: We are making BT work the same as before, everywhere. There is no "missing" or "changed behavior" stuff to log.

File: Custom-Audio-Driver/External-Audio-Device/OTDefaultAudioDevice.m (hunk 3 above)
Comment (IGitGotIt) [reply Custom-Audio-Driver/External-Audio-Device/OTDefaultAudioDevice.m:492]: done

File: Custom-Audio-Driver/External-Audio-Device/OTDefaultAudioDevice.m (hunk 4 above)
Comment (IGitGotIt) [reply Custom-Audio-Driver/External-Audio-Device/OTDefaultAudioDevice.m:497]: done

File: Custom-Audio-Driver/External-Audio-Device/OTDefaultAudioDevice.m (hunk 6 above)
Comment (IGitGotIt) [reply Custom-Audio-Driver/External-Audio-Device/OTDefaultAudioDevice.m:558]: We are making BT work the same as before, everywhere. There is no "missing" or "changed behavior" stuff to log.

File: Custom-Audio-Driver/External-Audio-Device/OTDefaultAudioDevice.m (hunk 5 above)
Comment (IGitGotIt) [reply Custom-Audio-Driver/External-Audio-Device/OTDefaultAudioDevice.m:506]: yep.

**Pull request merged**